lib/elixir/lib/module/types/descr.ex

```diff
@@ -75,6 +75,9 @@ defmodule Module.Types.Descr do
   def fun(), do: %{bitmap: @bit_fun}
   def list(type), do: list_descr(type, @empty_list, true)
   def non_empty_list(type, tail \\ @empty_list), do: list_descr(type, tail, false)
+  def improper_list(), do: improper_list(term(), term())
+  def improper_list(type, :term), do: list_descr(type, @not_list, false)
+  def improper_list(type, tail), do: list_descr(type, difference(tail, list(term())), false)
```

Comment:
We shouldn't need these helpers, as you would just pass a non list as second argument to `non_empty_list`. In particular, the internals were designed so you don't need to do compute the difference `difference(tail, list(term()))`.

I think what we want is:

    def list(type), do: list_descr(type, @empty_list, true)
    def non_empty_list(type), do: list_descr(type, @empty_list, false)
    def non_empty_maybe_improper_list(type, tail), do: list_descr(type, tail, false)

Then, when rendering, we would have:

    non_empty_list (proper, not empty)
    non_empty_maybe_improper_list (maybe improper, not empty)

    list (proper, maybe empty)
    maybe_improper_list (maybe improper, maybe empty)

WDYT?

Reply:
Also please remember to update `gradual-set-theoretic-types.md`

Reply:
> I think what we want is:

> def list(type), do: list_descr(type, @empty_list, true)
def non_empty_list(type), do: list_descr(type, @empty_list, false)
def non_empty_maybe_improper_list(type, tail), do: list_descr(type, tail, false)

OK, but wouldn't we need `improper_list(type, tail)`, at list for rendering, for the case where we know the last element cannot be `[]`?
e.g. `[1, 2, 3]` would be a valid (`non_empty_maybe_improper_list(integer(), binary())`) but not an `improper_list(integer(), binary())`. `[1, 2, 3 | "!"]` would be both.

Reply:
They are uncommon enough that I think saying maybe_improper_list is fine. Especially because once we have type variables and similar, we wouldn’t be able to say for sure which one it is.

Reply:
Plus we can always add later!

Reply:
Exactly!

Reply:
OK I think we're almost good, but I found something fishy in a test: [c61294f](https://github.com/elixir-lang/elixir/pull/14488/commits/c61294f92be8f11fb09e02d1d8fea1afce976408)

I won't have time to dive into it, but it might have deeper implications into our approach.